### PR TITLE
[using-webhooks] android-java

### PR DESCRIPTION
**Pull request by bg-stripe**
### Testing
- [ ] 4000 0000 0000 3220 -> shows native 3ds2 flow -> success
- [ ] 4000 0084 0000 1629 -> shows webview 3ds1 flow -> decline

### CODE REVIEW

File: using-webhooks/client/android-java/app/build.gradle [hunk 1]
```diff
@@ -0,0 +1,46 @@
+apply plugin: 'com.android.application'
+
+apply plugin: 'kotlin-android'
+
+apply plugin: 'kotlin-android-extensions'
+
+android {
+    compileSdkVersion 28
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/build.gradle]: 29

File: using-webhooks/client/android-java/app/build.gradle [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    defaultConfig {
+        applicationId "com.example.app"
+        minSdkVersion 21
+        targetSdkVersion 28
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/build.gradle]: 29

File: using-webhooks/client/android-java/app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    compileOptions {
+        sourceCompatibility 1.8
+        targetCompatibility 1.8
+    }
+}
+
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/build.gradle]: remove Kotlin dependency

File: using-webhooks/client/android-java/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    testImplementation 'junit:junit:4.12'
+    androidTestImplementation 'androidx.test:runner:1.1.0-alpha4'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0-alpha4'
+
+    // Stripe Android SDK
+    implementation 'com.stripe:stripe-android:11.1.3'
+
+    // OkHttp & GSON
+    implementation 'com.squareup.okhttp3:okhttp:4.2.0'
+    implementation 'com.google.code.gson:gson:2.8.5'
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/build.gradle]: update dependencies

File: using-webhooks/client/android-java/app/src/test/java/com/example/app/ExampleUnitTest.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test
+
+import org.junit.Assert.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/test/java/com/example/app/ExampleUnitTest.kt]: delete

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            PaymentMethodCreateParams params = cardInputWidget.getPaymentMethodCreateParams();
+            if (params != null) {
+                ConfirmPaymentIntentParams confirmParams = ConfirmPaymentIntentParams
+                        .createWithPaymentMethodCreateParams(params, paymentIntentClientSecret);
+                stripe.confirmPayment(this, confirmParams);
+            }
+        });
+    }
+
+    private void displayAlert(Activity activity, String title, String message, boolean restartDemo) {
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: add `@NonNull` annotations to activity, title, and message?

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                } else if (status == PaymentIntent.Status.RequiresPaymentMethod) {
+                    // Payment failed – allow retrying using a different payment method
+                    displayAlert(weakActivity.get(), "Payment failed", paymentIntent.getLastPaymentError().message, false);
+                }
+            }
+
+            @Override
+            public void onError(@NonNull Exception e) {
+                // Payment request failed – allow retrying using the same payment method
+                displayAlert(weakActivity.get(), "Error", e.toString(), false);
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: `weakActivity.get()` can be null

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            if (params != null) {
+                ConfirmPaymentIntentParams confirmParams = ConfirmPaymentIntentParams
+                        .createWithPaymentMethodCreateParams(params, paymentIntentClientSecret);
+                stripe.confirmPayment(this, confirmParams);
+            }
+        });
+    }
+
+    private void displayAlert(Activity activity, String title, String message, boolean restartDemo) {
+        runOnUiThread(() -> {
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: you shouldn't need to use `runOnUiThread` here, can you confirm that you do?

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        RequestBody body = RequestBody.create(json, mediaType);
+        Request request = new Request.Builder()
+                .url(backendUrl + "create-payment-intent")
+                .post(body)
+                .build();
+        httpClient.newCall(request)
+                .enqueue(new Callback() {
+                    @Override
+                    public void onFailure(Call call, IOException e) {
+                        runOnUiThread(() -> {
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java:82]: this should be running on the main thread so I don't think you need to do `runOnUiThread`

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                            Context applicationContext = getApplicationContext();
+                            Toast.makeText(applicationContext, "Error: " + e.toString(), Toast.LENGTH_LONG).show();
+                        });
+                    }
+
+                    @Override
+                    public void onResponse(Call call, final Response response) throws IOException {
+                        if (!response.isSuccessful()) {
+                            runOnUiThread(() -> {
```
Comment (mshafrir-stripe) [inline using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java:91]: this should be running on the main thread so I don't think you need to do `runOnUiThread`

**Review by mshafrir-stripe — CHANGES_REQUESTED** [review]

File: using-webhooks/client/android-java/app/build.gradle (hunk 4 above)
Comment (yuki-stripe) [reply using-webhooks/client/android-java/app/build.gradle]: @mshafrir-stripe ~How do I do that?~ alt-enter on the line and hit the suggestion.  That's cool.

File: using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java (hunk 8 above)
Comment (yuki-stripe) [reply using-webhooks/client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: thought I saw a crash or something, but I can't repro.  Removing.

**Pull request merged**